src/wx/widgets/wx/sdljoy.h

```diff
@@ -1,138 +1,102 @@
+//
+// After initilization, use PollJoystick() or PollAllJoysticks() for the
+// joysticks you wish to monitor.
+// The target window will receive EVT_SDLJOY events of type wxSDLJoyEvent.
 class wxSDLJoy : public wxTimer {
 public:
     wxSDLJoy();
-    // add another joystick to the list of polled sticks
-    // -1 == add all
-    // If joy > # of joysticks, it is ignored
-    // This will start polling if a valid joystick is selected
-    void Add(int8_t joy = -1);
-    // remove a joystick from the polled sticks
-    // -1 == remove all
-    // If joy > # of joysticks, it is ignored
-    // This will stop polling if all joysticks are disabled
-    void Remove(int8_t joy = -1);
-    // query if a stick is being polled
-    bool IsPolling(uint8_t joy) { return contains(joystate, joy); }
-
-    // true = currently rumbling, false = turn off rumbling
-    void SetRumble(bool do_rumble);
+    ~wxSDLJoy() override;
 
-    void Poll();
+    // Adds a joystick to the list of polled joysticks.
+    // If |index| is not an active index, the joystick will be added when it is
+    // connected.
+    // Otherwise, this will start polling if it was not previously started.
+    void PollJoystick(int index);
+
+    // Adds all joysticks to the list of polled joysticks.
+    // This will start polling.
+    void PollAllJoysticks();
+
+    // Removes all joysticks from the list of polled joysticks.
+    // This will stop polling.
+    void StopPolling();
 
-    virtual ~wxSDLJoy();
+    // Activates or deactivates rumble on active joysticks.
+    void SetRumble(bool activate_rumble);
+
+    // Polls active joysticks and empties the SDL event buffer.
+    void Poll();
 
 protected:
-    // used to continue rumbling on a timer
-    void Notify();
-    void ConnectController(uint8_t joy);
-    void RemapControllers();
-    void DisconnectController(wxSDLJoyState& dev);
-    void CreateAndSendEvent(unsigned short joy, unsigned short ctrl_type, unsigned short ctrl_idx, short ctrl_val, short prev_val);
+    // Utility method to simplify connecting a controller.
+    void ConnectController(int index);
+
+    // wxTimer implementation.
+    // Used to rumble on a timer.
+    void Notify() override;
 
     const uint8_t POLL_TIME_MS = 25;
 
 private:
-    std::unordered_map<uint8_t, wxSDLJoyState> joystate;
-    std::unordered_map<SDL_JoystickID, wxSDLJoyState*> instance_map;
-    bool add_all = false, rumbling = false;
+    // Map of SDL joystick ID to joystick state. Only contains active joysticks.
+    std::unordered_map<SDL_JoystickID, std::unique_ptr<wxSDLJoyState>> joystick_states_;
 
-    wxLongLong last_poll = wxGetUTCTimeMillis();
-};
-
-enum {
-    // The types of supported controls
-    // values are signed-16 for axis, 0/1 for button
-    // hat is bitmask NESW/URDL
-    WXSDLJOY_AXIS,
-    WXSDLJOY_HAT,
-    WXSDLJOY_BUTTON
-};
+    // Set of requested SDL joystick indexes.
+    std::unordered_set<int> requested_indexes_;
```

Comment:
This should be an unsigned type.

Reply:
I changed the API here so it's an int since this is an SDL index.